include/pybind11/detail/smart_holder_type_casters.h

```diff
@@ -698,6 +698,9 @@ struct smart_holder_type_caster<std::unique_ptr<T, D>> : smart_holder_type_caste
 
         return inst.release();
     }
+    static handle cast(const std::unique_ptr<T, D> &src, return_value_policy policy, handle parent) {
```

Comment:
Aha :-)
That makes everything clear immediately, thanks!
Initial thought: This is very similar to `static handle cast(const T *src, ...)`. We could mimic that implementation, or maybe even reuse? The easy case is that an instance exists already and we just incref as you suggest. Or there needs to be `return_value_policy::reference_internal` or an explicit `keep_alive`.
Is there a chance you could help with that implementation? It's unlikely that I can prioritize working on this myself anytime soon, unless I hit on a use case in PyCLIF (difficult to predict). But I will review & merge very quickly if you take the lead.
My revised suggestion for a separate new unit test name *iff* we cannot reuse `static handle cast(const T *src, ...)`: test_class_sh_const_unique_return
Keeping the tests separate helps minimizing merge issues (between master, drake, smart_holder) and is best practice IMO regardless.

Reply:
Done. Thanks for the hint to `cast(const T*)`. Regarding the unit test: I think we should have a unit test in any case.
Do you suggest to move the test into a new file, named `test_class_sh_const_unique_return.cpp` or to include a test named `test_class_sh_const_unique_return` into some existing file?

Reply:
Hm. I don't understand why CI fails, while locally it works fine. Need to investigate later...

Reply:
Oh, sorry to be ambiguous: I was thinking of `test_class_sh_const_unique_return.cpp` and `.py`.
I think I also want to add a few lines to `test_class_sh_basic.cpp` and `.py`, but that's a small enough effort I could squeeze that in myself.